### PR TITLE
Fix stack overflow caused by calls to List.map

### PR DESCRIPTION
**Pull request by zshipko**
I have been running into a stack overflow when building trees with >750,000 entries and passing them as an argument to `Store.set_tree`. This isn't exactly an ideal pattern but it should probably be supported.

It turns out that calls to `Stdlib.Map.bindings` are the source of the problem - I'm not sure if this PR is the best way to address the issue, so I'm open to any other suggestions. 

[Here](https://github.com/mirage/irmin/files/6073208/example.zip) is an example

### CODE REVIEW

**Comment by Ngoguey42** [comment]
I could reproduce the error thanks to your script, but the error doesn't come from `Map.bindings`, it comes from a `List.map` call in `Tree.export`. Replacing it with a `List.rev_map |> List.rev` seem to fix the problem. (https://github.com/mirage/irmin/blob/master/src/irmin/tree.ml#L1290)

If I'm not mistaken, `Map.bindings` works with a stack depth equal to the depth of the tree (log complexity). 
 (https://github.com/ocaml/ocaml/blob/trunk/stdlib/map.ml?ts=4#L489)

This bug is almost caught by [test_wide_nodes](https://github.com/mirage/irmin/blob/master/src/irmin-test/store.ml#L1407), we should update it to also commit the wide tree.

---

The stacktrace:
```
Fatal error: exception Stack overflow
Raised by primitive operation at Stdlib__map.Make.bindings_aux in file "map.ml", line 490, characters 42-73
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Irmin__Tree.Make.export.add_to_todo.(fun) in file "src/irmin/tree.ml", line 1295, characters 28-69
Called from Irmin__Tree.Make.export.loop in file "src/irmin/tree.ml", line 1353, characters 58-62
Called from Irmin__Tree.Make.export.(fun) in file "src/irmin/tree.ml", line 1356, characters 4-11
Called from Irmin__Store.Make.Commit.v.(fun) in file "src/irmin/store.ml", line 129, characters 21-54
Called from Irmin__Store.Make.update.(fun) in file "src/irmin/store.ml", line 737, characters 19-56
Called from Irmin__Store.Make.retry.aux in file "src/irmin/store.ml", line 646, characters 8-13
Called from Irmin__Store.Make.set_tree_exn in file "src/irmin/store.ml", line 759, characters 4-55
```

**Comment by zshipko** [comment]
Thanks for investigating! I am testing it out right now and it seems good.

I will update that test you mentioned as well.

**Comment by zshipko** [comment]
Updating that test shows there is a similar issue specific to `irmin-git`/`ocaml-git`:

```
[exception] Stack overflow
            Raised by primitive operation at Git__Tree.v.(fun) in file "src/git/tree.ml", line 128, characters 25-52
            Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
            Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
            Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
            Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
```

**Comment by dinosaure** [comment]
It seems that we have too many entries in that case for `ocaml-git` which uses `List.map` (not tail-rec version). I can propose something on my side to fix the bug. But the release will not appear asap (I need to check some others details).

**Comment by CraigFe** [comment]
For the Irmin case, I'd be in favour of shadowing `Stdlib.List.map` with a tail recursive version in `import.ml`. Such bugs are hard to spot and annoying to test for.

**Comment by zshipko** [comment]
@dinosaure: Sounds fine to me. I am most concerned with `irmin-pack` right now so I am able to continue, but it would definitely be nice to include a fix for irmin-git when possible.

@CraigFe Agreed, a general fix like that will probably save us lots of headaches - I will try that out.

**Comment by zshipko** [comment]
I added the tail recursive `map` implementation, still trying to understand how it affects performance though.

Do we want to wait for the `ocaml-git` fix to merge this? Or should I remove the failing test for now?

**Comment by dinosaure** [comment]
see mirage/ocaml-git#476 (a release is in the pipe)

**Comment by zshipko** [comment]
Great, that was very quick! Looks like we can wait on those changes to merge this.

I also realized that the wide trees test is being skipped on CI so it’s not an issue either way.

**Comment by CraigFe** [comment]
> I also realized that the wide nodes test is being skipped on CI so it’s not an issue either way.

I think we should be running the slow tests in CI, otherwise they're unlikely to get much use.

It's been in the back of my mind to fix that for a while, but we don't have anything tracking it yet.

**Comment by samoht** [comment]
There is still an issue with Git it seems:

```
ASSERT nb of contents folded over
+595155602us irmin.tree      [DEBUG] Tree.remove /foo/499999
+595155663us irmin.tree      [DEBUG] Tree.add
+595155677us irmin.tree      [DEBUG] Tree.add_tree /foo/499999
+595155696us irmin.tree      [DEBUG] Tree.hash
[exception] Stack overflow
            Raised by primitive operation at Stdlib__list.rev_append in file &quot;list.ml&quot;, line 58, characters 27-36
            Called from Stdlib__list.map in file &quot;list.ml&quot;, line 92, characters 32-39
            Called from Stdlib__list.map in file &quot;list.ml&quot;, line 92, characters 32-39
            Called from Stdlib__list.map in file &quot;list.ml&quot;, line 92, characters 32-39
            Called from Irmin__Tree.Make.Node.value_of_map.aux in file &quot;src/irmin/tree.ml&quot;, line 524, characters 22-49
            Called from Irmin__Tree.Make.hash in file &quot;src/irmin/tree.ml&quot;, line 1554, characters 23-36
            Called from Irmin__Store.Make.Tree.hash in file &quot;src/irmin/store.ml&quot;, line 89, characters 21-28
            Called from Irmin_test__Store.Make.test_wide_nodes.test.(fun) in file &quot;src/irmin-test/store.ml&quot;, line 1435, characters 15-29
            Called from Irmin_test__Common.Make_helpers.run.(fun) in file &quot;src/irmin-test/common.ml&quot;, line 156, characters 11-20
            Called from Lwt.Sequential_composition.bind.create_result_promise_and_callback_if_deferred.callback in file &quot;src/core/lwt.ml&quot;, line 1864, characters 23-26
            Re-raised at Irmin_test__Common.Make_helpers.run.(fun) in file &quot;src/irmin-test/common.ml&quot;, line 159, characters 36-72
            Called from Lwt.Sequential_composition.catch.create_result_promise_and_callback_if_deferred.callback in file &quot;src/core/lwt.ml&quot;, line 2045, characters 23-28
            Re-raised at Lwt.Miscellaneous.poll in file &quot;src/core/lwt.ml&quot;, line 3072, characters 20-29
            Called from Lwt_main.run.run_loop in file &quot;src/unix/lwt_main.ml&quot;, line 31, characters 10-20
            Called from Lwt_main.run in file &quot;src/unix/lwt_main.ml&quot;, line 118, characters 8-13
            Re-raised at Lwt_main.run in file &quot;src/unix/lwt_main.ml&quot;, line 124, characters 4-13
            Called from Alcotest_engine__Core.Make.protect_test.(fun) in file &quot;src/alcotest-engine/core.ml&quot;, line 277, characters 17-23
            Called from Alcotest_engine__Monad.Identity.catch in file &quot;src/alcotest-engine/monad.ml&quot;, line 34, characters 31-35


Logs saved to `/src/_build/default/_build/_tests/irmin-unix/GIT.025.output'.
```

**Comment by samoht** [comment]
Could you rebase on top of master (now that #1318 is merged) to make sure everything is ok with the Git backend?

**Comment by samoht** [comment]
also: could you add an entry in the changelog?

**Comment by zshipko** [comment]
Hm, not sure that fixed it - I will try to investigate.

**Comment by zshipko** [comment]
This is working now with the latest ocaml-git and encore releases. I also added and enabled the `Commit wide nodes` test which checks for this stack overflow issue.

It looks like there is still a segmentation fault on ppc64 on CI but I'm not quite sure how to debug that.

**Comment by zshipko** [comment]
I was able to get a ppc64 environment set up yesterday. The issue looks like it may be related to a mutex that is not being released somewhere but I haven't started looking into that too much yet - here is the stack trace when the program exists unexpectedly (I'm not able to get the segmentation fault like CI):

```
Thread 5 (Thread 0x7fffb6dff0f0 (LWP 15524)):
#0  0x00007ffff7c62a04 in __GI___select (nfds=<optimized out>, readfds=0x0, writefds=0x0, exceptfds=0x0, tim
eout=0x7fffb6dfe5b8) at ../sysdeps/unix/sysv/linux/select.c:41
#1  0x0000000100d0d0b0 in caml_thread_tick (arg=<optimized out>) at st_posix.h:388
#2  0x00007ffff7ed9370 in start_thread (arg=0x7fffb6dff0f0) at pthread_create.c:463
#3  0x00007ffff7c716d4 in clone () at ../sysdeps/unix/sysv/linux/powerpc/powerpc64/clone.S:82

Thread 3 (Thread 0x7ffff682f0f0 (LWP 15466)):
#0  0x00007ffff7ee3084 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x10140864c
 <pool_condition+44>) at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x1014085e8 <pool_mutex>, cond=0x101408620 <po
ol_condition>) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=0x101408620 <pool_condition>, mutex=0x1014085e8 <pool_mutex>) at pthread_cond_
wait.c:638
#3  0x0000000100cffb48 in lwt_unix_condition_wait (condition=<optimized out>, mutex=<optimized out>) at lwt_
unix_stubs.c:225
#4  0x0000000100cfff98 in worker_loop (data=<optimized out>) at lwt_unix_stubs.c--Type <RET> for more, q to 
quit, c to continue without paging--c
:937
#5  0x00007ffff7ed9370 in start_thread (arg=0x7ffff682f0f0) at pthread_create.c:463
#6  0x00007ffff7c716d4 in clone () at ../sysdeps/unix/sysv/linux/powerpc/powerpc64/clone.S:82

Thread 2 (Thread 0x7ffff703f0f0 (LWP 15465)):
#0  0x00007ffff7ee3084 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x10140864c <pool_condition+44>) at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x1014085e8 <pool_mutex>, cond=0x101408620 <pool_condition>) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=0x101408620 <pool_condition>, mutex=0x1014085e8 <pool_mutex>) at pthread_cond_wait.c:638
#3  0x0000000100cffb48 in lwt_unix_condition_wait (condition=<optimized out>, mutex=<optimized out>) at lwt_unix_stubs.c:225
#4  0x0000000100cfff98 in worker_loop (data=<optimized out>) at lwt_unix_stubs.c:937
#5  0x00007ffff7ed9370 in start_thread (arg=0x7ffff703f0f0) at pthread_create.c:463
#6  0x00007ffff7c716d4 in clone () at ../sysdeps/unix/sysv/linux/powerpc/powerpc64/clone.S:82

Thread 1 (Thread 0x7ffff7ff4680 (LWP 15460)):
#0  0x00007ffff7b719d8 in __GI_exit (status=2) at exit.c:138
#1  0x0000000100d4af3c in caml_fatal_uncaught_exception (exn=<optimized out>) at printexc.c:160
#2  0x0000000100d26920 in caml_startup (argv=<optimized out>) at startup_nat.c:170
#3  caml_main (argv=<optimized out>) at startup_nat.c:175
#4  0x00000001005d265c in main (argc=<optimized out>, argv=<optimized out>) at main.c:41
```

**Comment by zshipko** [comment]
Alright, this is finally passing now! 

The the issue was `List.concat` in `irmin-pack` - let me know if you see any other issues, if not this should be good to merge.

**Comment by samoht** [comment]
Can you use List.rev_concat instead? It could be costly to traverse the second list

**Review by Ngoguey42 — APPROVED** [review]

File: src/irmin/import.ml [hunk 1]
```diff
@@ -51,6 +51,15 @@ module List = struct
       | h :: t -> (aux [@tailcall]) (fun t' -> acc (f h :: t')) t
     in
     aux (fun x -> x) l
+
+  let concat l =
+    let rec aux acc curr l =
+      match (curr, l) with
+      | [], [] -> List.rev acc
+      | [], h :: t -> (aux [@tailcall]) acc h t
```
Comment (samoht) [inline src/irmin/import.ml:59]: I think it could be a good idea to have a special case fo `[], [l]` (and return `List.rev_append acc l`)

**Comment by CraigFe** [comment]
> This is working now with the latest ocaml-git and encore releases. I also added and enabled the `Commit wide nodes` test which checks for this stack overflow issue and modified the irmin-test to allow irmin-http tests to skip it since it takes way too long.

Out of interest, how long do these tests take?

**Comment by zshipko** [comment]
> Out of interest, how long do these tests take?

@CraigFe Good question... For most of the backends it seems to add a few seconds, but I just pushed a change to make slow tests opt-in instead of opt-out and disabled them for `irmin-pack` because it added several minutes (which is probably something else worth looking in to)

Without any slow tests: `0m33.170s`
Slow tests enabled for irmin-pack: `5m25.581s`
Current state (enabled for mem, git, fs, chunk): `0m41.376s`
